### DISCUSSION TITLE
How can I provide a specific context to a specific test?

### DISCUSSION THREAD

**Discussion by artgocode** [cucumber-rs/cucumber#228] [Q&A] [answered]
Hi!
First of all I want to thank you for the effort you put into this project. This crate is awesome!

I am now facing ambiguity issues when running my tests.
I need help to wrap my head around on how can I provide a specific context to my tests?
Under "contex" I mean the `.rs` files with `given`, `when`, `then` expressions defined for a specific `.feature` files.

I am developing a backend api based on actix_web framework.
My workspace contains two members as for now. Application crate with library and my `test` crate witch contains all cucumber tests/features.
My `test` workspace member has next structure:

```
src/_
    |_lib.rs            // mod helpers; mod context;
    |_helpers.rs        // here I put all common structs/functions/assertion helpers for all context files, no "given", "when", "then" expressions here
    |_context/
        |_api/
           |_admin/
               |_managing_currencies_context.rs // here I define all "given", "when", "then" matches for "managing_currencies.feature" 
               |_managing_locales_context.rs // here I define all "given", "when", "then" matches for "managing_locales.feature"
tests/
    |_features/
    |   |_currency/
    |   |   |_managing_currencies/
    |   |       |_adding_currency.feature
    |   |       |_removing_currency.feature
    |   |_locale/
    |       |_managing_locales/
    |           |_adding_locale.feature
    |           |_removing_locale.feature
    |_currency/
    |   |_adding_currency.rs
    |   |_removing_currency.rs
    |_locale/
        |_adding_locale.rs
```

`adding_locale.rs` contains code:

```
use cucumber::WorldInit;
use test::helpers::TestApp;

#[tokio::main]
async fn main() {
    TestApp::run("features/locale/managing_locales/adding_locale.feature").await;
}
```

`adding_currency.rs` contains code:

```
use cucumber::WorldInit;
use test::helpers::TestApp;

#[tokio::main]
async fn main() {
    TestApp::run("features/currency/managing_currencies/adding_currency.feature").await;
}

```

`Cargo.toml` for `test` crate:

```
[package]
***
[lib]
name = "test"

[[test]]
name = "adding_locale"                
harness = false                       
path = "tests/locale/adding_locale.rs"

[[test]]
name = "adding_currency"
harness = false
path = "tests/currancy/adding_currency.rs"
```
`TestApp` definition from `helpers.rs` file:

```
#[derive(Debug, WorldInit)]
pub struct TestApp {
    pub api_client: ApiClient,
    pub db_pool: hryven::DbPool,
}

#[async_trait(?Send)]
impl World for TestApp {
    // We do require some error type.
    type Error = Infallible;

    async fn new() -> Result<Self, Infallible> {
        Ok(spawn_app().await) // spawn_app() bootstraps the application instance
    }
}
```

I am now trying to figure out, how can I provide a `managing_currencies_context.rs` context only to `adding_currency.rs` and `removing_currency.rs` tests?
Or maybe there is another way to structure my tests?
I tried to find some examples of a large test code bases, but failed to.

**Comment by ilslv** [comment, marked as answer]
@artgocode first of all, I hope you are doing well. Слава Україні!

---

> Or maybe there is another way to structure my tests?

We intend for users of this crate to structure tests in a quite particular way, mostly because Rust doesn't support custom test runners very well yet.  I think good example of `cucumber` usage would be [`pagebreak`](https://github.com/CloudCannon/pagebreak) or [`algonaut`](https://github.com/manuelmauro/algonaut). 

But to sum up a bit:

### Folders structure

```
src/_
    | main.rs        // Actual application logic to test. In your case, probably `actix_web` backend.
tests/
    | cucumber.rs   // `World` definition. Code for it leaves in your `helpers.rs` for now.
    |_features/
    |   | *         // Everything here may be left in tact.  
                    // But now `cucumber.rs` file and it's definitions inside are actually visible.
Cargo.toml
```

### `Cargo.toml`

Add those lines:

```
[[test]]
name = "cucumber"
harness = false   # Allows Cucumber to print output instead of libtest
```

And then just run `cargo test --test cucumber` and everything should work fine. 

For further reading I suggest [Quickstart](https://cucumber-rs.github.io/cucumber/main/quickstart.html) and [Modules organization](https://cucumber-rs.github.io/cucumber/main/writing/modules.html#modules-organization) book chapters.

**Reply by artgocode** [reply]
Thank you very much @ilslv for clarifications you gave.

I am curious about how steps are being discovered by cucumber?

**Reply by ilslv** [reply]
@artgocode we are using [`inventory`](https://github.com/dtolnay/inventory) crate, which links to a specific section of the binary and involves life-before-main. [`linkme`](https://github.com/dtolnay/linkme) can also achieve similar results by doing linking shenanigans, but uses alternative approach and doesn't involved in life-before-main.

There was a push to bring this kind of functionality inside Rust language itself, but it hasn't resulted in a RFC yet afaik.
